nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/analyzeflow/ruleimpl/DisallowComponentType.java

```diff
@@ -0,0 +1,84 @@
+ * limitations under the License.
+ */
+package org.apache.nifi.analyzeflow.ruleimpl;
+
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.flow.VersionedComponent;
+import org.apache.nifi.flow.VersionedExtensionComponent;
+import org.apache.nifi.flowanalysis.AbstractFlowAnalysisRule;
+import org.apache.nifi.flowanalysis.ComponentAnalysisResult;
+import org.apache.nifi.flowanalysis.FlowAnalysisRuleContext;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+
+@Tags({"component", "processor", "controller service", "type"})
+@CapabilityDescription("Produces rule violations for each component (i.e. processors or controller services) of a given type.")
+public class DisallowComponentType extends AbstractFlowAnalysisRule {
+    public static final PropertyDescriptor COMPONENT_TYPE = new PropertyDescriptor.Builder()
+            .name("component-type")
+            .displayName("Component Type")
+            .description("Components of the given type will produce a rule violation (i.e. they shouldn't exist).")
+            .required(true)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .defaultValue(null)
+            .build();
+
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.add(COMPONENT_TYPE);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @Override
+    public Collection<ComponentAnalysisResult> analyzeComponent(VersionedComponent component, FlowAnalysisRuleContext context) {
+        Collection<ComponentAnalysisResult> results = new HashSet<>();
+
+        String componentType = context.getProperty(COMPONENT_TYPE).getValue();
+
+        if (component instanceof VersionedExtensionComponent) {
+            VersionedExtensionComponent versionedExtensionComponent = (VersionedExtensionComponent) component;
+
+            String encounteredComponentType = versionedExtensionComponent.getType();
+            String encounteredSimpleComponentType = encounteredComponentType.substring(encounteredComponentType.lastIndexOf(".") + 1);
+
+            if (encounteredComponentType.equals(componentType) || encounteredSimpleComponentType.equals(componentType)) {
+                ComponentAnalysisResult result = ComponentAnalysisResult.newResult(
+                        "default",
+                        "'" + componentType + "' is not allowed!"
```

Comment:
Exclamation marks should not be used in messages.
```suggestion
                        "'" + componentType + "' is not allowed"
```